onnx/onnx.in.proto

```diff
@@ -196,6 +202,125 @@ message NodeProto {
+//      initializer list.
+//   2. Prepare other inputs (usually are user-provided tensors) which are not
+//      initializers.
+//   3. Bind all inputs to the "algorithm" and execute "algorithm". The
+//      "algorithm" is usually the computation graph of one training iteration
+//      (i.e., code to process one batch of data).
+//   4. Assign selected outputs of "algorithm" to some specified initializers.
+message TrainingInfoProto {
+  // Training-specific initializers, for example, accumulated squared gradient
+  // in ADAGRAD optimizer, momentum, or number of executed training iterations.
+  // The values can only be inputs of "algorithm" defined below.
+  repeated TensorProto initializer = 1;
+
+  // Training inputs, for example, labels in classification problems.
+  // Those values can only be inputs of "algorithm" defined below.
+  repeated ValueInfoProto input = 2;
+
+  // Training outputs, for example, new tensor values computed by the
+  // specified training algorithm. Those values can only be selected outputs of
+  // "algorithms".
+  repeated ValueInfoProto output = 3;
+
+  // The training algorithm. Given required inputs, it computes outputs which
+  // can be used to update initializers. In general, this field contains
+  // loss node, gradient node, optimizer node, and calls to the inference
+  // function. The field "algorithm.node" can reference functions defined in
+  // the field "ModelProto.function".
+  //
+  // The field MUST be present.
+  optional FunctionProto algorithm = 4;
+
+  // Gradient-based training is usually an iterative procedure. In one gradient
+  // descent iteration, we apply
+  //
+  // x = x - r * g
+  //
+  // where "x" is the optimized tensor, "r" stands for learning rate, and "g" is
+  // gradient of "x" with respect to a chosen loss. To avoid adding assignments
+  // into the training graph (ONNX does not have assignment operator yet), the
+  // user need to save "y = x - r * g" into a TrainingProto. To tell runtime
+  // that "y" should be assigned to "x", the field "update_binding" should
+  // contain a pair of strings, "y" (key of StringStringEntryProto) and "x"
+  // (value of StringStringEntryProto). Similarly, for a neural network which
+  // contains multiple tensors, there will be multiple key-value pairs in
+  // "update_binding". In general, "update_binding" can assign any output of
+  // "algorithm" to any initializer in either "ModelProto.graph.initializer" or
+  // "initializer". For not breaking single-static-assignment, one tensor can
+  // only be assigned up to once; it implies that the values in "update_binding"
+  // must be unique.
+  repeated StringStringEntryProto update_binding = 7;
+
+  // A graph that describes how to compute the initial tensors upon starting
+  // the training process. Each initializer graph has zero inputs and one
+  // output, which acts as the initialized tensor. The initialization graph
```

Comment:
I think the "each initializer graph" sentence is a hangover from a previous iteration, although capuring "zero inputs" is still valuable.

Reply:
Thanks for catching this. I changed it to multiple outputs. Similar to training algorithm, the assignment of initial values are defined by `update_binding` now. In other words, this spec treats model initialization as a kind of `training`.